File: java/sodium/src/sodium/EventSink.java
package sodium;

import java.util.List;

public class EventSink<A> extends Event<A> {
    public EventSink() {}

	public void send(final A a) {
		Transaction.run(new Handler<Transaction>() {
			public void run(Transaction trans) { send(trans, a); }
		});
	}

    void send(Transaction trans, A a) {
        if (firings.isEmpty())
            trans.last(new Runnable() {
            	public void run() { firings.clear(); }
            });
        firings.add(a);
        
        @SuppressWarnings("unchecked")
		List<TransactionHandler<A>> listeners = (List<TransactionHandler<A>>)this.listeners.clone();
    	for (TransactionHandler<A> action : listeners) {
    		try {
                action.run(trans, a);
    		}
    		catch (Throwable t) {
    		    t.printStackTrace();
    		}
    	}
    }
}

File: c#/Sodium/src/Sodium/EventSink.cs
using System;
using System.Collections.Generic;

namespace Sodium
{

  public class EventSink<TA> : Event<TA>
  {
    public EventSink() { }

    public void Send(TA a)
    {
      Transaction.Run(new Handler<Transaction>
      {
        Run = trans => Send(trans, a)
      });
    }

    public void Send(Transaction trans, TA a)
    {
      if (Firings.IsEmpty())
        trans.Last(new Runnable
        {
          Run = () => Firings.Clear()
        });
      Firings.Add(a);

      var listeners = (List<TransactionHandler<TA>>)Listeners.Clone();
      foreach (TransactionHandler<TA> action in listeners)
      {
        try
        {
          action.Run(trans, a);
        }
        catch (Exception e)
        {
          Console.WriteLine(e);
        }
      }
    }
  }
}